File: Week04/problem3.jl
using CSV
using DataFrames


function pd_return_calculate(prices::Vector, method::String="arithmetic")
    price_change_percent = diff(prices) ./ prices[1:end-1]
    if method == "arithmetic"
        return price_change_percent .- 1
    elseif method == "log"
        return log.(price_change_percent)
    end
end

# Load data and calculate returns
prices = CSV.File("/Users/oliver/Desktop/FinTech 545/FinTech-545-Spring2023-main/Week04/DailyPrices.csv"; dateformat="yyy-mm-dd") |> DataFrame
portfolios = CSV.File("/Users/oliver/Desktop/FinTech 545/FinTech-545-Spring2023-main/Week04/Project/portfolio.csv") |> DataFrame
println(size(prices))
println(size(portfolios))
returns = pd_return_calculate(prices[:, 2], "log")

# Combine the portfolios to get a total one and append it to the end for easier calculation
total_holdings = combine(groupby(portfolios, :Stock), :Holding => sum => :Holding)
total_holdings[!, :Portfolio] .= "Total"
append!(portfolios, total_holdings)


function shapiro_test(data; alpha=0.05)
    test_stat, p = shapiro(data)
    if p > alpha
        return 1
    else
        return 0
    end
end

# Load data and calculate returns
prices = CSV.File("DailyPrices.csv"; dateformat="yyy-mm-dd") |> DataFrame
portfolios = CSV.File("portfolio.csv") |> DataFrame
returns = pd_return_calculate(prices[:, 2], "log")

# Determine if the returns are normally distributed using Shapiro-Wilks test
for (portfolio_index, portfolio) in groupby(portfolios, :Portfolio)
    portfolio_returns = returns[:, portfolio.Stock]
    num_normal = apply(portfolio_returns, 2, shapiro_test) |> sum
    percentage_normal = num_normal / size(portfolio_returns, 2) * 100
    println(string(percentage_normal) * "%")
end

current_prices = DataFrame(Price=prices[end, :])

for (portfolio_index, portfolio) in groupby(portfolios, :Portfolio)
    portfolio = portfolio |> setindex!(:Stock)
    portfolio = join(portfolio, current_prices[portfolio[!, :Stock], :])

    current_values = portfolio[!, :Holding] .* portfolio[!, :Price]
    portfolio_value = sum(current_values)

    sim_returns = returns[!, portfolio[!, :Stock]]
    sim_prices = (1 .+ sim_returns) .* portfolio[!, :Price]'
    sim_values = sim_prices * portfolio[!, :Holding]

    historic_var = calculate_var(sim_values, portfolio_value)
    println("Portfolio $(portfolio_index): $(historic_var)")
end

File: Week05/problem3.py
from Library import calculate_ewcov, historical_ES,readfile,Calculate_his,Calculate_norm
import pandas as pd
from scipy.stats import norm, t
import numpy as np

def calculate_returns(price_series, method='arithmetic'):
    changes = price_series.pct_change()[1:]
    if method == 'arithmetic':
        returns = changes
    elif method == 'log':
        returns = np.log(1 +changes)

    return returns

# load in data and calculate returns
prices = pd.read_csv("/Users/oliver/Desktop/FinTech 545/Week05/Project/DailyPrices.csv", parse_dates=[0], index_col=0)
portfolios = pd.read_csv("/Users/oliver/Desktop/FinTech 545/Week05/Project/portfolio.csv")
returns = calculate_returns(prices)

total = portfolios.groupby('Stock').sum('Holding')
total['Portfolio'] = 'Total'
total = total.reset_index()
portfolios = portfolios.append(total)

t_params = {}
sim_data = []
for col in returns:
  stock_returns = returns[col]

  stock_returns -= stock_returns.mean()
  result = t.fit(stock_returns, method="MLE")
  df, loc, scale = result
  t_params[col] = [df, loc, scale]
  sim_data.append(t(df, loc, scale).rvs(10000))
sim_data = np.array(sim_data)

simulated_returns = pd.DataFrame(columns=returns.columns, data=sim_data.T)


current_prices = pd.DataFrame({"Price":prices.iloc[-1]})

#Fit a Generalized T model 

for name, portfolio in portfolios.groupby('Portfolio'):
    portfolio = portfolio.set_index('Stock')
    portfolio = portfolio.join(current_prices.loc[portfolio.index])

    sim_returns = simulated_returns[portfolio.index]
    sim_prices_change = sim_returns * portfolio['Price'].T
    sim_values_change = sim_prices_change @ portfolio['Holding']

    var = Calculate_his(sim_values_change)
    es = historical_ES(sim_values_change)
    print(f" {name} " + "VaR: " + str(var))
    print(f" {name} " + "ES: " + str(es))
    print()
     

